ui/client/reducers/graph/utils.ts

```diff
@@ -1,30 +1,51 @@
+import {cloneDeep, map, omit, reject, without, zipWith, curryRight} from "lodash"
+import {Layout, NodePosition, NodesWithPositions} from "../../actions/nk"
+import ProcessUtils from "../../common/ProcessUtils"
 import {ExpressionLang} from "../../components/graph/node-modal/editors/expression/types"
-import {GraphState} from "./types"
-import {NodeType, NodeId, Process, GroupType, Edge, EdgeType, ProcessDefinitionData} from "../../types"
 import NodeUtils from "../../components/graph/NodeUtils"
-import {map, zipWith, omit, cloneDeep, reject} from "lodash"
-import {Layout, NodesWithPositions, NodePosition} from "../../actions/nk"
-import ProcessUtils from "../../common/ProcessUtils"
+import {Edge, EdgeType, GroupType, NodeId, NodeType, Process, ProcessDefinitionData} from "../../types"
+import {GraphState} from "./types"
 
-function isBetween(id1: NodeId, id2: NodeId): (e: Edge) => boolean {
-  return ({from, to}) => from == id1 && to == id2 || from == id2 && to == id1
+function isConnectedTo(...nodeIds: NodeId[]) {
+  return (edge: Edge) => {
+    return nodeIds.find(id => edge.from === id || edge.to === id)
+  }
 }
 
-function canGroup(state: GraphState, newNode: NodeType | GroupType): boolean {
-  const {groupingState, processToDisplay: {edges}} = state
-  const isGroup = NodeUtils.nodeIsGroup(newNode)
-  return !isGroup && groupingState.length === 0 || !!groupingState.find(id => edges.find(isBetween(id, newNode.id)))
+function getSelectedNodes(state: GraphState): NodeType[] {
+  const {processToDisplay, selectionState = []} = state
+  return NodeUtils.nodesFromProcess(processToDisplay).filter(n => selectionState.includes(n.id))
 }
 
-export function displayOrGroup(state: GraphState, node: NodeType, readonly = false): GraphState {
-  if (state.groupingState) {
-    return {
-      ...state,
-      groupingState: canGroup(state, node) ?
-        state.groupingState.concat(node.id) :
-        state.groupingState,
-    }
+export function getSelectedGroups(state: GraphState): GroupType[] {
+  return getSelectedNodes(state).filter(NodeUtils.nodeIsGroup)
+}
+
+export function canGroupSelection(state: GraphState): boolean {
```

Comment:
Maybe some unit test would be nice for this?

Reply:
coverage from cypress, do you really need unit?

<img width="548" alt="Zrzut ekranu 2021-05-5 o 12 53 47" src="https://user-images.githubusercontent.com/965924/117130884-f1523000-ada0-11eb-9100-4a76c3439d60.png">

Reply:
maybe not *that* really, but I think cypress coverage is misleading in this case - lines are ok, but I don't think current tests capture all cases...